### PR TITLE
Hash ring

### PR DESCRIPTION
**Pull request by achille-roussel**
Here's the implementation of a high performance consistent hash ring. The API is much simpler that what hashring libraries usually provide, it doesn't allow adding or removing servers, instead the intent is that the program will recreate the entire ring if the list of servers has changed.

Tests and Benchmarks:
```
$ go test -v -run TestHash -bench Hash -benchmem
=== RUN   TestHash
=== RUN   TestHash/ensures_the_hash_function_behaves_like_the_standard_fnv1-a
=== RUN   TestHash/ensures_the_hash_function_behaves_like_the_standard_fnv1-a/#00
=== RUN   TestHash/ensures_the_hash_function_behaves_like_the_standard_fnv1-a/A
=== RUN   TestHash/ensures_the_hash_function_behaves_like_the_standard_fnv1-a/Hello_World!
=== RUN   TestHash/ensures_the_hash_function_behaves_like_the_standard_fnv1-a/DAB45194-42CC-4106-AB9F-2447FA4D35C2
--- PASS: TestHash (0.00s)
    --- PASS: TestHash/ensures_the_hash_function_behaves_like_the_standard_fnv1-a (0.00s)
        --- PASS: TestHash/ensures_the_hash_function_behaves_like_the_standard_fnv1-a/#00 (0.00s)
        --- PASS: TestHash/ensures_the_hash_function_behaves_like_the_standard_fnv1-a/A (0.00s)
        --- PASS: TestHash/ensures_the_hash_function_behaves_like_the_standard_fnv1-a/Hello_World! (0.00s)
        --- PASS: TestHash/ensures_the_hash_function_behaves_like_the_standard_fnv1-a/DAB45194-42CC-4106-AB9F-2447FA4D35C2 (0.00s)
=== RUN   TestHashRing
--- PASS: TestHashRing (0.00s)
	hashing_test.go:107: going from 2 to 3 servers redistributed ~30% of the keys (301/1000)
	hashing_test.go:107: going from 3 to 4 servers redistributed ~18% of the keys (183/1000)
	hashing_test.go:107: going from 4 to 5 servers redistributed ~23% of the keys (235/1000)
	hashing_test.go:107: going from 3 to 5 servers redistributed ~41% of the keys (418/1000)
BenchmarkHash/standard_hash_function-4         	20000000	       106 ns/op	 338.92 MB/s	      56 B/op	       2 allocs/op
BenchmarkHash/hash_function-4                  	20000000	        72.6 ns/op	 495.72 MB/s	       0 B/op	       0 allocs/op
BenchmarkHashRing-4                            	20000000	       108 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/segmentio/redis-go	6.093s
```

Part of #12

### CODE REVIEW

File: hashing.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+// keys to server addresses.
+type hashRing []hashNode
+
+type hashNode struct {
+	addr string
+	hash uint64
+}
+
+const (
+	hashRingReplication = 40
```
Comment (Pryz) [inline hashing.go]: Why 40 ?

Comment (achille-roussel) [inline hashing.go]: It gave good results, that's the value another hash ring package I found was using as well.

**Comment by achille-roussel** [comment]
Just changed the hashing algorithm to use jody, pretty cool perf:
- with fnv-1a:
```
$ go test -v -run TestHash -bench Hash -benchmem
=== RUN   TestHashRing
--- PASS: TestHashRing (0.00s)
	hashring_test.go:57: going from 2 to 3 servers redistributed ~30% of the keys (303/1000)
	hashring_test.go:57: going from 3 to 4 servers redistributed ~15% of the keys (154/1000)
	hashring_test.go:57: going from 4 to 5 servers redistributed ~18% of the keys (188/1000)
	hashring_test.go:57: going from 3 to 5 servers redistributed ~33% of the keys (337/1000)
BenchmarkHashRing-4   	20000000	        93.4 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/segmentio/redis-go	1.992s
```
- with jody:
```
$ go test -v -run TestHash -bench Hash -benchmem
=== RUN   TestHashRing
--- PASS: TestHashRing (0.00s)
	hashring_test.go:57: going from 2 to 3 servers redistributed ~14% of the keys (144/1000)
	hashring_test.go:57: going from 3 to 4 servers redistributed ~14% of the keys (147/1000)
	hashring_test.go:57: going from 4 to 5 servers redistributed ~6% of the keys (61/1000)
	hashring_test.go:57: going from 3 to 5 servers redistributed ~20% of the keys (208/1000)
BenchmarkHashRing-4   	20000000	        59.3 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/segmentio/redis-go	1.268s
```

**Comment by achille-roussel** [comment]
Thanks for the review!

**Pull request merged**